doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,321 @@
+on the `StorageController` trait, I would like to add a new method:
+
+```
+pub async fn monotonic_appender(&self, id: GlobalId) -> CollectionAppender;
+```
+
+This new method would return a struct called `CollectionAppender`, which essentially acts as a
+`oneshot::Sender<Vec<(Row, Diff)>>`. It's a `Send`-able and non-`Clone`-able struct that has one
+method `async fn append(self, ...)` where it takes ownership of `self` so it can only be used once.
+This would be implemented as part of [`collection_mgmt`](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs)
+and would approximately be structed like:
+
+```
+struct CollectionAppender {
+  id: GlobalId,
+  tx: mpsc::Sender<(GlobalId, Vec<(Row, Diff)>)>,
+}
+```
+
+In addition to the `CollectionAppender`, we'll update the `CollectionManager` to receive a
+`oneshot::Sender<Result<(), StorageError>>` that it can use to optionally notify any listeners when
+and if the updates were successfully committed.
+
+Adding the `CollectionAppender` introduces a possible correctness issue though:
+
+1. Create a `CollectionAppender` for `GlobalId(A)`.
+2. Drop source associated with `GlobalId(A)`.
+3. Try to append data.
+
+To protect against this we can change the `CollectionManager` to return an error via the
+`oneshot::Sender` if the [provided GlobalId no longer exists](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller/collection_mgmt.rs#L83).
+
+Adding this new API gets us two important properties that the existing `append` API does not have:
+
+1. The ability to send updates to the `StorageController` from a thread other than the main
+   `Coordinator` thread.
+2. Allows the `StorageController` to pick the timestamp for the update, instead of the
+   `Coordinator`, which aligns with our long term goals if having a storage owned "push source" in
+   the future.
+
+### Request Validation
+[request_validation]: #request_validation
+
+Every `WEBHOOK` source will be open to the entire internet, as such we need to include some sort of
+validation so we can verify that requests are legitimate. In general webhooks are validated by
+HMAC-ing the request body using SHA256, and validating the result with a signature provided in the
+request headers. An issue though is everyone does this just a little bit different:
+
+* Segment signature = HMAC of just the body of the request
+* Buildkite signature = HMAC of "#{timestamp}.#{body}"
+* GitHub signature = "body=" + HMAC of {body}
+
+As such we'll need to support some custom logic for validation. What we can do is support a
+`VALIDATE USING` statement that accepts only a single scalar expression. This expression will
+be provided the request `headers` as `map[text] -> text`, and the `body` as `text` _regardless of
+whether or not the source has `INCLUDE HEADERS` or what `FORMAT BODY` is specified._ For example:
+
+```
```

Comment:
Allowing _any_ kind of custom logic makes me nervous, but do we have a good idea of what kind of operations we can comfortably support right now (and what kinds of validations we want to do on the expression)? Is it non-viable support one hardcoded method right now (e.g. what Segment does, because that's simplest), and expand to more "recipes" as suggested in future work?

Reply:
What about custom logic makes you nervous? In the future we will probably send all this work off to clusterd so envd will be protected from anything complicated. For now, this is good enough for the target ship date.

Reply:
How does that differ from the current HTTP API?

Reply:
Materialize is in the business of executing user-submitted code. That's fundamentally what `SELECT` and `CREATE INDEX` do. Indeed it is easy to shoot yourself in the foot (see `WITH MUTUALLY RECURSIVE` in particular), and the suggestions in the thread about making webhooks easy to monitor and disable are great points, but we needn't stress about `VALIDATE USING` allowing an arbitrary expression. As @sploiselle, the existing HTTP API already lets you execute arbitrary SQL.

> Is it non-viable to support _only_ one hardcoded method right now (e.g. what Segment does, because that's simplest), and expand to more "recipes" as suggested in future work?

Yeah, the MVP needs to ship support for more than just one webhook provider, and survey of five webhook providers we did found five different signature methods. I also want to avoid _needing_ to ship a custom recipe for every webhook provider that someone wants to rig up; that means potentially weeks of lead time to onboard a user who wants to use a webhok provider.